taf/tests/test_updater/test_update_library/test_validation_and_sync.py

```diff
@@ -0,0 +1,214 @@
+        origin_auth_repo,
+        client_dir,
+    )
+
+
+@pytest.mark.parametrize(
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_auth_repo_not_in_sync_partial(origin_auth_repo, client_dir):
+    """
+    Clients auth repo has been manually updated, just like in 1.
+    The update should result in a partial update.
+    Clients auth repo should be reverted to the last valid commit,
+    last validated commits set to that value,
+    target repositories updated up to commits listed in auth repo's last validated commits
+    """
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(update_expiration_dates)
+    setup_manager.add_task(add_valid_target_commits)
+    setup_manager.add_task(
+        revert_last_validated_commit, kwargs={"client_dir": client_dir}
+    )
+    setup_manager.execute_tasks()
+
+    assert_repositories_updated(client_dir, origin_auth_repo)
+
+
+@pytest.mark.parametrize(
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_target_repo_not_in_sync_invalid_commits(origin_auth_repo, client_dir):
+    """
+    Same as 2, but target repositories contain invalid commits following commits which were manually pulled using git pull.
+    Expect all repositories to be updated up to the last validated commits and invalid commits to be removed
+    """
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(add_valid_unauthenticated_commits)
+    setup_manager.add_task(update_expiration_dates)
+    setup_manager.add_task(
+        sync_target_repos_with_remote,
+        kwargs={"origin_auth_repo": origin_auth_repo, "client_dir": client_dir},
+    )
+    setup_manager.add_task(
+        revert_last_validated_commit, kwargs={"client_dir": client_dir}
+    )
+    setup_manager.execute_tasks()
+
+    assert_repositories_updated(client_dir, origin_auth_repo)
+
+
+@pytest.mark.parametrize(
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_mixed_target_repo_states(origin_auth_repo, client_dir):
+    """
+    Client's auth repo is not on last validated commit, but target commits are all over the place.
```

Comment:
I probably wanted to say here that auth repo is on last validated commit, but we could test both cases